File: engine/src/main/java/de/hanno/hpengine/engine/model/QuadVertexBuffer.java
package de.hanno.hpengine.engine.model;

import de.hanno.hpengine.engine.graphics.GpuContext;

import javax.vecmath.Vector2f;
import java.util.EnumSet;

public class QuadVertexBuffer extends VertexBuffer {

	public QuadVertexBuffer(GpuContext gpuContext, boolean fullscreen) {
		super(gpuContext, EnumSet.of(DataChannels.POSITION3, DataChannels.TEXCOORD), getPositionsAndTexCoords(fullscreen));
	}
	public QuadVertexBuffer(GpuContext gpuContext, Vector2f leftBottom, Vector2f rightUpper) {
		super(gpuContext, EnumSet.of(DataChannels.POSITION3, DataChannels.TEXCOORD), getPositionsAndTexCoords(leftBottom, rightUpper));
	}

	static float[] getPositionsAndTexCoords(boolean fullscreen) {
		if (fullscreen) {
			return new float[] {
					-1.0f, -1.0f, 0.0f,   0f, 0f,
					1.0f, -1.0f, 0.0f,    1f, 0f,
					-1.0f,  1.0f, 0.0f,   0f,  1.0f,
					-1.0f,  1.0f, 0.0f,   0f,  1.0f,
					1.0f, -1.0f, 0.0f,    1.0f, 0f,
					1.0f,  1.0f, 0.0f,    1.0f,  1.0f
			};
		} else {
//			return new float[] {
//					-1.0f, -1.0f, 0.0f, 0f, 0f,
//					0f, -1.0f, 0.0f,    1f, 0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					0f, -1.0f, 0.0f,    1.0f, 0f,
//					0f,  0f, 0.0f,      1.0f, 1.0f
//			};
			return getPositionsAndTexCoords(new Vector2f(-1, -1), new Vector2f(0, 0));
		}
	}
	static float[] getPositionsAndTexCoords(Vector2f leftBottom, Vector2f rightUpper) {
		return new float[] {
				leftBottom.x, leftBottom.y, 0.0f,   0f, 0f,
				rightUpper.x, leftBottom.y, 0.0f,    1f, 0f,
				leftBottom.x,  rightUpper.y, 0.0f,   0f,  1.0f,
				leftBottom.x,  rightUpper.y, 0.0f,   0f,  1.0f,
				rightUpper.x, leftBottom.y, 0.0f,    1.0f, 0f,
				rightUpper.x,  rightUpper.y, 0.0f,    1.0f,  1.0f
		};
	}
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/vertexbuffer/QuadVertexBuffer.kt
package de.hanno.hpengine.engine.vertexbuffer

import de.hanno.hpengine.engine.graphics.GpuContext
import java.util.*
import javax.vecmath.Vector2f

class QuadVertexBuffer : VertexBuffer {
    constructor(gpuContext: GpuContext<*>?, fullscreen: Boolean) : super(
        gpuContext!!,
        EnumSet.of<DataChannels>(DataChannels.POSITION3, DataChannels.TEXCOORD),
        getPositionsAndTexCoords(fullscreen)
    ) {
    }

    constructor(gpuContext: GpuContext<*>?, leftBottom: Vector2f, rightUpper: Vector2f) : super(
        gpuContext!!,
        EnumSet.of<DataChannels>(DataChannels.POSITION3, DataChannels.TEXCOORD),
        getPositionsAndTexCoords(leftBottom, rightUpper)
    ) {
    }

    companion object {
        fun getPositionsAndTexCoords(fullscreen: Boolean): FloatArray {
            return if (fullscreen) {
                floatArrayOf(
                    -1.0f, -1.0f, 0.0f, 0f, 0f,
                    1.0f, -1.0f, 0.0f, 1f, 0f,
                    -1.0f, 1.0f, 0.0f, 0f, 1.0f,
                    -1.0f, 1.0f, 0.0f, 0f, 1.0f,
                    1.0f, -1.0f, 0.0f, 1.0f, 0f,
                    1.0f, 1.0f, 0.0f, 1.0f, 1.0f
                )
            } else {
//			return new float[] {
//					-1.0f, -1.0f, 0.0f, 0f, 0f,
//					0f, -1.0f, 0.0f,    1f, 0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					0f, -1.0f, 0.0f,    1.0f, 0f,
//					0f,  0f, 0.0f,      1.0f, 1.0f
//			};
                getPositionsAndTexCoords(
                    Vector2f(-1f, -1f),
                    Vector2f(0f, 0f)
                )
            }
        }

        fun getPositionsAndTexCoords(leftBottom: Vector2f, rightUpper: Vector2f): FloatArray {
            return floatArrayOf(
                leftBottom.x, leftBottom.y, 0.0f, 0f, 0f,
                rightUpper.x, leftBottom.y, 0.0f, 1f, 0f,
                leftBottom.x, rightUpper.y, 0.0f, 0f, 1.0f,
                leftBottom.x, rightUpper.y, 0.0f, 0f, 1.0f,
                rightUpper.x, leftBottom.y, 0.0f, 1.0f, 0f,
                rightUpper.x, rightUpper.y, 0.0f, 1.0f, 1.0f
            )
        }
    }
}